test_runner/src/main/kotlin/ftl/doctor/Doctor.kt

```diff
@@ -5,43 +5,49 @@ import com.google.common.annotations.VisibleForTesting
 import ftl.args.AndroidArgsCompanion
 import ftl.args.ArgsHelper
 import ftl.args.IArgs
-import ftl.args.yml.DEVICES_NODE
-import ftl.args.yml.GCLOUD_NODE
 import ftl.args.yml.MODEL_NODE
 import ftl.args.yml.VERSION_NODE
 import ftl.args.yml.devicesNode
 import ftl.args.yml.notValidDevices
 import ftl.config.loadAndroidConfig
 import ftl.config.loadIosConfig
+import ftl.domain.DoctorErrors
+import ftl.domain.plus
 import ftl.run.exception.FlankConfigurationError
 import ftl.util.loadFile
 import java.io.Reader
-import java.lang.StringBuilder
 import java.nio.file.Path
 
-fun validateYaml(args: IArgs.ICompanion, data: Path) =
-    if (!data.toFile().exists()) "Skipping yaml validation. No file at path $data"
+fun validateYaml(args: IArgs.ICompanion, data: Path): DoctorErrors =
+    if (!data.toFile().exists()) DoctorErrors(parsingErrors = listOf("Skipping yaml validation. No file at path $data"))
     else validateYaml(args, loadFile(data)) + preloadConfiguration(data, args is AndroidArgsCompanion)
 
 @VisibleForTesting
-internal fun validateYaml(args: IArgs.ICompanion, data: Reader) =
+internal fun validateYaml(args: IArgs.ICompanion, data: Reader): DoctorErrors =
     runCatching { ArgsHelper.yamlMapper.readTree(data) }
-        .onFailure { return it.message?.replace(System.lineSeparator(), "\n") ?: "Unknown error when parsing tree" }
+        .onFailure {
+            return DoctorErrors(
+                parsingErrors = listOf(
+                    it.message?.replace(System.lineSeparator(), "\n") ?: "Unknown error when parsing tree"
+                )
+            )
+        }
         .getOrNull()
-        ?.run { validateYamlKeys(args).plus(validateDevices()) }
-        .orEmpty()
+        ?.run { validateYamlKeys(args) }
+        ?: DoctorErrors.EMPTY
 
-private fun JsonNode.validateYamlKeys(args: IArgs.ICompanion) = StringBuilder().apply {
-    append(validateTopLevelKeys(args))
-    args.validArgs.forEach { (topLevelKey, validArgsKeys) ->
-        append(validateNestedKeys(topLevelKey, validArgsKeys))
-    }
-}.toString()
+private fun JsonNode.validateYamlKeys(args: IArgs.ICompanion) = DoctorErrors(
+    topLevelUnknownKeys = validateTopLevelKeys(args),
+    nestedUnknownKeys = args.validArgs.map { (topLevelKey, validArgsKeys) ->
+        (topLevelKey to validateNestedKeys(topLevelKey, validArgsKeys))
+    }.filter { it.second.isNotEmpty() }.toMap(),
+    invalidDevices = validateDevices().orEmpty()
+)
 
-private fun JsonNode.validateTopLevelKeys(args: IArgs.ICompanion) =
+private fun JsonNode.validateTopLevelKeys(args: IArgs.ICompanion): List<String> =
     (parseArgs().keys - args.validArgs.keys)
         .takeIf { it.isNotEmpty() }
-        ?.let { unknownKeys -> "Unknown top level keys: $unknownKeys\n" }
+        ?.map { it }
```

Comment:
what is the purpose of this `map`?

Reply:
That is an excellent question...and I don't know 😅 will be removed, Thanks!

Reply:
We expecting ```List``` as result of function but ```(parseArgs().keys - args.validArgs.keys)``` returns ```Set```. I will change this map to ```.toList()```